miner/miner.go

```diff
@@ -192,3 +204,138 @@ func (miner *Miner) SetEtherbase(addr common.Address) {
 func (self *Miner) SubscribePendingLogs(ch chan<- []*types.Log) event.Subscription {
 	return self.worker.pendingLogsFeed.Subscribe(ch)
 }
+
+// commitmentCacheSaved will check to see if the last commitment cache entry is saved.
+// If not, it will search for it and save it.
+func (miner *Miner) commitmentCacheSaved(istEngine consensus.Istanbul) bool {
+	// Subscribe to new block notifications
+	newBlockCh := make(chan core.ChainEvent)
+	bc := miner.eth.BlockChain()
+	newBlockSub := bc.SubscribeChainEvent(newBlockCh)
+	defer newBlockSub.Unsubscribe()
+
+	// getCurrentBlockAndState
+	currentBlock := bc.CurrentBlock()
+	currentHeader := currentBlock.Header()
+	currentState, err := bc.StateAt(currentBlock.Root())
+	if err != nil {
+		log.Error("Error in retrieving state", "block hash", currentHeader.Hash(), "error", err)
+		return false
+	}
+
+	// Check to see if we already have the commitment cache
+	lastCommitment, err := random.GetLastCommitment(miner.coinbase, currentHeader, currentState)
+	if err != nil {
+		log.Error("Error in retrieving last commitment", "error", err)
+		return false
+	}
+
+	// If there was no previous commitment from this validator, then return true.
+	if (lastCommitment == common.Hash{}) {
+		return true
+	}
+
+	if (rawdb.ReadRandomCommitmentCache(miner.db, lastCommitment) != common.Hash{}) {
+		return true
+	}
+
+	// goroutine communication channels and waitgroup
```

Comment:
this second part of the function looks weird...

it reads like:
```
firstBlock = race(
  getMostRecentFromCoinbaseInHistory()
  getFirstBlockFronCoinbaeInFuture()
)
generateAndSaveCommitement(firstBlock)

```

now, because of the race, this doesn't look to be deterministic, hence weird.

Reply:
How would you suggest this is changed?

Both `getters` of the race is necessary.

Reply:
nvm, i see some more of you comments below.

Reply:
Do we need to wait for new blocks at this point? During sync start, mining gets stopped with `Mining aborted due to sync`, but as it reads in the new blocks that it doesn't have, it will check if it should save the randomness. Once it's done syncing or fails to sync, it runs this function prior to starting mining again.

Reply:
I have the same doubt as @trianglesphere . I understood that when this function is called we are "already synced" since that's the reason of the `update()` function. (but i remember it's a tricky one, since it only triggers once)

And it would seems that even if that's not the case, the code in blockchain would still handle it, since it reacts on new blocks

Reply:
This handles the case of when the the partition head block (made up of the node and it's peers when the sync `downloader.DoneEvent` is emitted) is behind a block that the node authored (let's call this the "future authored block"). Note that this probably is a very rare scenario.

When this scenario happens, two things can happen.

1) The node's reverse iteration is completed before the "future authored block" is synced.  The node will then start istanbul.  Even if this partition's aggregate head block is behind other validators, this is already handled via the istanbul engine.
2) The "future authored block" is synced before the reverse iteration is completed.  In this case, the syncing of that block will stop the reverse iteration since it has become irrelevant, and then istanbul will be started.

So waiting for new blocks is basically to handle case 2.  Having said all that, that case is probably going to happen very infrequently, and the added benefit (not having to wait for the reverse iteration to complete) seems marginal.  So seems like it's not worth the added complexity.  WDYT?

Reply:
So thinking about what we store, it's a lookup of `commitment -> block hash` where the block hash is mangled to create the commitment. I think that we have multiple of these commitments in the eth db. When looking up the last commitment made, we check a smart contract, so I think that order does not matter here, only that we do have the last commitment saved in the level db.

It's true that we want to bail out of reverse iteration as an optimization if we find a "future authored block", but shouldn't the future block record its commitment when we process it.

Reply:
@trianglesphere - Ya, you're correct. It's unnecessarily recalculating the cache entry when it gets the "future authored block".

WDYT of the proposal of removing that handling of new block logic altogether. It does add alot of complex logic (which means higher chance for bugs) for minor benefit?

Reply:
That sounds good to me. Seem like it's rare enough and doesn't make it incorrect.

Reply:
+1